data-prepper-core/src/main/java/com/amazon/dataprepper/parser/PipelineParser.java

```diff
@@ -91,15 +94,16 @@ private void buildPipelineFromConfiguration(
             final PluginSetting sourceSetting = pipelineConfiguration.getSourcePluginSetting();
             final Optional<Source> pipelineSource = getSourceIfPipelineType(pipelineName, sourceSetting,
                     pipelineMap, pipelineConfigurationMap);
-            final Source source = pipelineSource.orElseGet(() -> SourceFactory.newSource(sourceSetting));
+            final Source source = pipelineSource.orElseGet(() ->
+                    pluginFactory.loadPlugin(Source.class, sourceSetting));
 
             LOG.info("Building buffer for the pipeline [{}]", pipelineName);
-            final Buffer buffer = BufferFactory.newBuffer(pipelineConfiguration.getBufferPluginSetting());
+            final Buffer buffer = pluginFactory.loadPlugin(Buffer.class, pipelineConfiguration.getBufferPluginSetting());
```

Comment:
I moved away from the enum approach and toward a class-based approach so that we can support plugins for other class types. One extension I'd like to make is to make some of our common configurations pluggable. For example, several plugins load SSL certificates from different locations. This can be pluggable so that 1) The code is shared; and 2) New locations can be added without having to change the original code.

The enum will not provide the necessary flexibility.